paddle/fluid/framework/save_load_util.cc

```diff
@@ -0,0 +1,372 @@
+      used_tensor_message += " " + tensor_temp.first;
+    }
+
+    LOG(ERROR) << used_tensor_message;
+  }
+
+  return true;
+}
+
+bool SaveTensorToDisk(const std::string& file_name,
+                      const std::map<std::string, Tensor*>& map_tensor) {
+  MkDirRecursively(DirName(file_name).c_str());
+
+  std::ofstream fout(file_name, std::ios::binary);
+  if (!fout) {
+    PADDLE_THROW("File open error. Can not open file [%s]", file_name);
+  }
+
+  // first 256 byte for reserve for fulture upgrade
+  char* kReserveBuffer = new char[model_file_reserve_size];
+  fout.write(kReserveBuffer, sizeof(char) * model_file_reserve_size);
+  delete[] kReserveBuffer;
+
+  fout.write(tensor_number_mark.c_str(),
+             sizeof(char) * tensor_number_mark.size());
+  size_t tensor_number = map_tensor.size();
+  fout.write(reinterpret_cast<const char*>(&tensor_number),
+             sizeof(tensor_number));
+
+  for (auto& itera : map_tensor) {
+    // first save tensor name
+    fout.write(tensor_name_mark.c_str(),
+               sizeof(char) * tensor_name_mark.size());
+    size_t name_length = itera.first.size();
+    fout.write(reinterpret_cast<const char*>(&name_length),
+               sizeof(name_length));
+    fout.write(itera.first.c_str(), sizeof(char) * name_length);
+    // write tensor version
+    constexpr uint32_t version = 0;
+    fout.write(reinterpret_cast<const char*>(&version), sizeof(version));
+
+    // the 2nd field, tensor description
+    // int32_t  size
+    // void*    protobuf message
+    auto tensor = itera.second;
+
+    proto::VarType::TensorDesc desc;
+    desc.set_data_type(tensor->type());
+    auto dims = framework::vectorize(tensor->dims());
+    auto* pb_dims = desc.mutable_dims();
+    pb_dims->Resize(static_cast<int>(dims.size()), 0);
+    std::copy(dims.begin(), dims.end(), pb_dims->begin());
+    int32_t size = desc.ByteSize();
+    fout.write(reinterpret_cast<const char*>(&size), sizeof(size));
+    auto out = desc.SerializeAsString();
+    fout.write(out.data(), size);
+
+    // save tensor
+    uint64_t data_size =
+        tensor->numel() * framework::SizeOfType(tensor->type());
+    auto* data_ptr = tensor->data<void>();
+    if (platform::is_gpu_place(tensor->place())) {
+#ifdef PADDLE_WITH_CUDA
+      framework::Tensor temp;
+      TensorCopySync(*tensor, platform::CPUPlace(), &temp);
+      data_ptr = temp.data<void>();
+#else
+      PADDLE_THROW(
+          "Tensor is in CUDA device, but paddle not compile with CUDA, this "
+          "shoul'd not happen");
+#endif
+    }
+    fout.write(static_cast<const char*>(data_ptr),
+               static_cast<std::streamsize>(data_size));
+  }
+
+  if (!fout) {
+    PADDLE_THROW("Model save failed, data write to model file [%s] error",
+                 file_name);
+  }
+
+  fout.close();
+
+  return true;
+}
+
+bool LoadTensorFromDisk(
+    const std::string& file_name,
+    std::map<std::string, std::shared_ptr<Tensor>>& map_tensor) {  // NOLINT
```

Comment:
maybe use pointer instead of reference is better

Reply:
result need to return